File: Fractals/EscapeTimeFractals/Mandelbrot/mandelbrot.py
import numpy as np
import matplotlib.pyplot as plt

# Mandelbrot Set function
def mandelbrot(c, max_iter):
    z = 0
    n = 0
    while abs(z) <= 2 and n < max_iter:
        z = z*z + c
        n += 1
    return n

# Function to generate Mandelbrot fractal
def mandelbrot_set(xmin, xmax, ymin, ymax, width, height, max_iter):
    r1 = np.linspace(xmin, xmax, width)
    r2 = np.linspace(ymin, ymax, height)
    fractal = np.empty((width, height))

    for i in range(width):
        for j in range(height):
            c = complex(r1[i], r2[j])
            fractal[i, j] = mandelbrot(c, max_iter)
    
    return fractal

# Parameters for fractal generation
xmin, xmax, ymin, ymax = -2.5, 1.5, -2.0, 2.0
width, height = 800, 800
max_iter = 256

# Generate the Mandelbrot fractal
fractal = mandelbrot_set(xmin, xmax, ymin, ymax, width, height, max_iter)

# Plot the result using Matplotlib
plt.imshow(fractal.T, cmap='hot', extent=[xmin, xmax, ymin, ymax])
plt.colorbar()
plt.title("Mandelbrot Set")
plt.show()

File: Fractals/EscapeTimeFractals/Mandelbrot/mandelbrot.go
package main

import (
	"image"
	"image/color"
	"image/png"
	"math/cmplx"
	"os"
)

// Mandelbrot function returns the number of iterations before a point escapes
func mandelbrot(c complex128, maxIter int) int {
	var z complex128
	for n := 0; n < maxIter; n++ {
		if cmplx.Abs(z) > 2 {
			return n
		}
		z = z*z + c
	}
	return maxIter
}

// Generate Mandelbrot set and save as PNG image
func generateMandelbrot(width, height, maxIter int, xmin, xmax, ymin, ymax float64, filename string) {
	img := image.NewRGBA(image.Rect(0, 0, width, height))

	for px := 0; px < width; px++ {
		for py := 0; py < height; py++ {
			x := float64(px)/float64(width)*(xmax-xmin) + xmin
			y := float64(py)/float64(height)*(ymax-ymin) + ymin
			c := complex(x, y)

			// Color based on the number of iterations
			m := mandelbrot(c, maxIter)
			colorVal := uint8(255 - (m * 255 / maxIter))
			col := color.RGBA{colorVal, colorVal, 255 - colorVal, 255}
			img.Set(px, py, col)
		}
	}

	// Save to file
	f, _ := os.Create(filename)
	defer f.Close()
	png.Encode(f, img)
}

func main() {
	// Image parameters
	width, height := 800, 800
	maxIter := 256
	xmin, xmax := -2.5, 1.5
	ymin, ymax := -2.0, 2.0

	// Generate and save the Mandelbrot set
	generateMandelbrot(width, height, maxIter, xmin, xmax, ymin, ymax, "mandelbrot.png")
}
